endpoint/verify.go

```diff
@@ -34,17 +39,47 @@ func NewVerifyEndpoint(
 			return
 		}
 
-		if !strings.HasPrefix(authHeader[0], "Bearer ") {
-			w.WriteHeader(http.StatusUnauthorized)
-			return
+		var key, token string
+		if strings.HasPrefix(authHeader[0], "Key ") {
+			key = strings.TrimPrefix(authHeader[0], "Key ")
+			var ok bool
+			if token, ok = tCache.FetchToken(r.Context(), key); !ok {
+				results := regexp.MustCompile(`^([^|]+)\|([^|]+)$`).FindStringSubmatch(key)
+				if len(results) == 0 {
+					w.WriteHeader(http.StatusUnauthorized)
+					return
+				}
+				clientID := results[1]
+				clientSecret := results[2]
+
+				var err error
+				token, err = apiKeys(clientID, clientSecret).GetToken(r.Context())
```

Comment:
Not sure if I'm completely following, but why wouldn't the calling service use their client id & secret to generate api keys, which they pass here to create a jwt?  Probably fine with https and all, but just seemed a little strange to be passing the client id and secret over the request.

Reply:
I guess the idea is to reduce the toil of having to do that song and dance on the server side, which makes sense!

Reply:
@bipol So just to make sure we're on the same page, there is no pre-existing notion of an "API key": JWTs are the primitive for all things access in our Auth0 setup. However, one of Auth0's many pathways for getting a JWT is this two-piece "client credentials" thing, so the idea is that by sticking those two pieces together into a string, we can jerry rig our own implementation of "API key", where the key is a set of client credentials.

The reason we ultimately want a JWT is because the key is opaque/doesn't tell us anything about who this is. So to make sure that it's a valid key and get information like scopes or a user ID, we ask Auth0 to convert the credentials into an access token. (We're not actually _using_ any scopes for anything, but if we did, we'd need to get that info from a JWT, not a key).

Reply:
Ah yeah, my description up there doesn't make any sense. What I meant to say is - our services use the client id and client secret to create the jwt, which we pass to you here, not that they retrieve and api key which obtains a jwt.

If we're trying to avoid toil, this definitely makes sense.  It did look odd to me that we are passing a client id and client secret around (just since I usually see that being kept by the service to obtain some form of key, since the id and secret allows you to create them as you see fit) - which made me pause here.

Reply:
I did find some prior art for auth0 in a thread from 2017 here https://community.auth0.com/t/how-to-implement-api-keys-using-auth0/6296/2 as well supporting this sort of scheme

Reply:
Yeah and I definitely agree that passing the secret with the request is not just strange but "risky" in that it if you grab that secret you can generate as many JWTs as you want forever, although I think it's slightly less risky than a non-expiring JWT, since we can always delete the client and kill your next refresh.

So as things grow, we'll want to make sure our internal clients have only the powers they need, so that we can feel comfortable passing these secrets around often, just like API keys

Reply:
Also, I'm not planning to merge this at least until Monday's meeting, and I threw it together in ~an hour and am very open to alternative suggestions!

Reply:
Ooooh good looking out! The idea of not requiring a handshake/caching the tokens might still be a little out there, but at least the underlying security principle has the support of the community